python/phase-2.md

```diff
@@ -0,0 +1,830 @@
+        pass
+
+    # Include same typed overloads as WorkflowHandle
+    async def signal(self, signal: Any, *args: Any) -> None:
+        pass
+
+    async def result(self) -> WorkflowReturnType:
+        pass
+
+    async def cancel(self) -> None:
+        pass
+
+# Overload for no-param workflow
+# TODO(cretz): Fix when https://github.com/python/mypy/issues/11833 fixed
+@overload
+async def start_child_workflow(
+    self,
+    workflow: Callable[[WorkflowType], Awaitable[WorkflowReturnType]],
+    /,
+    *,
+    # ...
+) -> ChildWorkflowHandle[WorkflowType, WorkflowReturnType]:
+    pass
+
+# Overload for single-param workflow
+# TODO(cretz): Fix when https://github.com/python/mypy/issues/11833 fixed
+@overload
+async def start_child_workflow(
+    self,
+    workflow: Callable[
+        [WorkflowType, LocalParamType], Awaitable[WorkflowReturnType]
+    ],
+    arg: LocalParamType,
+    /,
+    *,
+    # ...
+) -> ChildWorkflowHandle[WorkflowType, WorkflowReturnType]:
+    pass
+
+# Overload for string-name workflow
+@overload
+async def start_child_workflow(
+    self,
+    workflow: str,
+    *args: Any,
+    # ...
+) -> ChildWorkflowHandle[Any, Any]:
+    pass
+
+async def start_child_workflow(
+    self,
+    workflow: Any,
+    *args: Any,
+    id: str,
+    task_queue: Optional[str] = None,
+    cancellation_type: ChildWorkflowCancellationType = ChildWorkflowCancellationType.WAIT_CANCELLATION_COMPLETED,
+    parent_close_policy: ParentClosePolicy = ParentClosePolicy.TERMINATE,
+    execution_timeout: Optional[timedelta] = None,
+    run_timeout: Optional[timedelta] = None,
+    task_timeout: Optional[timedelta] = None,
+    id_reuse_policy: temporalio.common.WorkflowIDReusePolicy = temporalio.common.WorkflowIDReusePolicy.ALLOW_DUPLICATE,
+    retry_policy: Optional[temporalio.common.RetryPolicy] = None,
+    cron_schedule: str = "",
+    memo: Optional[Mapping[str, Any]] = None,
+    search_attributes: Optional[Mapping[str, Any]] = None,
+    header: Optional[Mapping[str, Any]] = None,
+) -> ChildWorkflowHandle[Any, Any]:
+    # Do stuff
+```
+
+Notes:
+
+* `ChildWorkflowHandle.signal` will have the same typed overloads that `WorkflowHandle.signal` does
+* `ChildWorkflowHandle.cancel` is present to simplify cancellation. See "temporalio.workflow Scoped Cancellation" for
+  more discussion.
+* `start_child` was considered instead of `start_child_workflow`, but this seems clearer. Can be convinced otherwise.
+
+#### `temporalio.workflow.execute_child_workflow()`
+
+Shortcut for `temporalio.workflow.start_child_workflow()` + `handle.result()`
+
+#### `temporalio.workflow.ChildWorkflowConfig`
+
+`TypedDict` of `start_child_workflow`/`execute_child_workflow` configuration options.
+
+#### `temporalio.workflow.continue_as_new()`
+
+```python
+# Overload for self (cannot type args)
+@overload
+def continue_as_new(
+    *args: Any,
+    # ... (excludes "workflow" kwarg)
+) -> NoReturn:
+    pass
```

Comment:
It's just too bad this isn't typed.

Could the workflow.run decorator shadow this definition to use the appropriate types so that it works out when called within workflow implementations?

Reply:
I can't think of a clean way to do this at compile time. TS hasn't solve this either, they also don't implicitly type the continue-as-new for self.

The only somewhat-reasonable way I could think of doing this would be to require the function passed in for every continue-as-new, even the self one.

This is one of the few costs of not having an explicitly generic typed context that we pass around everywhere and put all of these methods on (but of course the benefits of not having a Python user manually type and pass around some generic context everywhere are much larger).

Reply:
I would be fine with just eliminating this and requiring you to pass the workflow type. Don't feel super strongly about it though.